### ISSUE TITLE
DFU fails on Android 8 and 8.1 after "Reading DFU version number..." but succeeds on Android version < 8

### ISSUE THREAD

**Issue by beegee-tokyo** [NordicSemiconductor/Android-DFU-Library#173]
I use a simplified version of the nrfToolbox sources (removed unused features) and Android-DFU-Library V 1.9.0
The DFUActivity of the nRFTooldbox sources is basically **_unchanged_** I only replaced the deprecated preference values with the default ones. See code fragment below.

Bootloader/Softdevice is from [Adafruit nRF52 Arduino](https://github.com/adafruit/Adafruit_nRF52_Arduino), version used is 0.9.3

ZIP file for update is identical for all tests on Android 7.1.1, Android 8.0 and Android 8.1

If I run the app on Android 7 or older the DFU update works as expected, but on Android 8.0 (Huawei Mediapad M5) and 8.1 (Samsung Galaxy J7 core) the DFU update fails after device is forced into bootloader, reconnected and then trying to read the DFU version number. See Android catlog below. On Android 8 and 8.1 it seems the DFU service is too slow and the device reboots into the application while the DFU service still tries to read the DFU version.
On Android older than 7 I see in the timestamp ~90 milliseconds from requesting the DFU version to receiving the DFU version
On Android 8.1 I see in the timestamps ~ 5 seconds from requesting the DFU version to read error and disconnect.

Logcat shows in case of failure:
``` LOG
10:30:42.917  I/DfuImpl: Reading DFU version number...
10:30:47.139  E/DfuImpl: Characteristic read error: 133
10:30:47.140  D/BluetoothGatt: onClientConnectionState() - status=8 clientIf=7 device=E0:22:0C:0E:98:BD
10:30:47.140  W/DfuBaseService: Target device disconnected with status: 8
10:30:47.141  E/DfuBaseService: Unable to read version number: device disconnected
```
Any idea?

Code used to start DFU service:
``` JAVA
		final DfuServiceInitiator starter = new DfuServiceInitiator(mSelectedDevice.getAddress())
				.setDeviceName(mSelectedDevice.getName())
				.setKeepBond(true)
				.setForceDfu(true)
				.setPacketsReceiptNotificationsEnabled(true)
				.setPacketsReceiptNotificationsValue(DfuServiceInitiator.DEFAULT_PRN_VALUE)
				.setUnsafeExperimentalButtonlessServiceInSecureDfuEnabled(true)
				.setForeground(true);
		if (mFileType == DfuService.TYPE_AUTO) {
			starter.setZip(mFileStreamUri, mFilePath);
			if (mScope != null)
				starter.setScope(mScope);
		} else {
			starter.setBinOrHex(mFileType, mFileStreamUri, mFilePath).setInitFile(mInitFileStreamUri, mInitFilePath);
		}
		starter.start(this, DfuService.class);
```
Catlog from Android 7.1.1 (successful OTA update)
``` LOG
10:42:14.609  I/DfuBaseService: DFU service created. Version: 1.9.0
10:42:14.767  I/DfuBaseService: Connecting to the device...
10:42:14.769  D/BluetoothGatt: connect() - device: E0:22:0C:0E:98:BD, auto: false
10:42:14.769  D/BluetoothGatt: registerApp()
10:42:14.770  D/BluetoothGatt: registerApp() - UUID=b77baf40-0d60-401d-8d3c-6e5d06de6850
10:42:14.772  D/BluetoothGatt: onClientRegistered() - status=0 clientIf=6
10:42:16.997  D/BluetoothGatt: onClientConnectionState() - status=0 clientIf=6 device=E0:22:0C:0E:98:BD
10:42:16.999  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_CONNECTED
10:42:17.001  I/DfuBaseService: Connected to GATT server
10:42:17.006  D/BluetoothGatt: discoverServices() - device: E0:22:0C:0E:98:BD
10:42:17.010  I/DfuBaseService: Attempting to start service discovery... succeed
10:42:18.268  D/BluetoothGatt: onSearchComplete() = Device=E0:22:0C:0E:98:BD Status=0
10:42:18.269  I/DfuBaseService: Services discovered
10:42:19.284  I/DfuImpl: Reading DFU version number...
10:42:19.364  W/BluetoothGatt: onCharacteristicRead() - Device=E0:22:0C:0E:98:BD handle=21 Status=0
10:42:19.367  I/DfuImpl: Version number read: 0.1 -> Application with Legacy buttonless update from SDK 7.0 or newer
10:42:19.371  W/DfuImpl: Application with legacy buttonless update found
10:42:19.372  I/DfuImpl: Enabling notifications...
10:42:19.376  D/BluetoothGatt: setCharacteristicNotification() - uuid: 00001531-1212-efde-1523-785feabcd123 enable: true
10:42:19.874  I/art: Do partial code cache collection, code=231KB, data=200KB
10:42:19.875  I/art: After code cache collection, code=229KB, data=198KB
10:42:19.875  I/art: Increasing code cache capacity to 1024KB
10:42:20.486  I/DfuImpl: Sending Start DFU command (Op Code = 1, Upload Mode = 4)
10:42:40.575  D/BluetoothGatt: onClientConnectionState() - status=8 clientIf=6 device=E0:22:0C:0E:98:BD
10:42:40.575  W/DfuBaseService: Target device disconnected with status: 8
10:42:40.581  D/BluetoothGatt: refresh() - device: E0:22:0C:0E:98:BD
10:42:40.589  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_DISCONNECTED
10:42:40.593  I/DfuBaseService: Refreshing result: true
10:42:40.594  I/DfuBaseService: Cleaning up...
10:42:40.596  D/BluetoothGatt: close()
10:42:40.596  D/BluetoothGatt: unregisterApp() - mClientIf=6
10:42:40.599  I/DfuImpl: Starting service that will connect to the DFU bootloader
10:42:42.675  I/DfuBaseService: Connecting to the device...
10:42:42.679  D/BluetoothGatt: connect() - device: E0:22:0C:0E:98:BD, auto: false
10:42:42.679  D/BluetoothGatt: registerApp()
10:42:42.680  D/BluetoothGatt: registerApp() - UUID=f46d4775-1ba9-4c17-b9e1-c5b16858d6fd
10:42:42.683  D/BluetoothGatt: onClientRegistered() - status=0 clientIf=6
10:42:47.158  D/BluetoothGatt: onClientConnectionState() - status=0 clientIf=6 device=E0:22:0C:0E:98:BD
10:42:47.158  I/DfuBaseService: Connected to GATT server
10:42:47.160  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_CONNECTED
10:42:47.162  D/BluetoothGatt: discoverServices() - device: E0:22:0C:0E:98:BD
10:42:47.165  I/DfuBaseService: Attempting to start service discovery... succeed
10:42:47.792  D/BluetoothGatt: onSearchComplete() = Device=E0:22:0C:0E:98:BD Status=0
10:42:47.792  I/DfuBaseService: Services discovered
10:42:48.808  I/DfuImpl: Reading DFU version number...
10:42:48.899  W/BluetoothGatt: onCharacteristicRead() - Device=E0:22:0C:0E:98:BD handle=21 Status=0
10:42:48.901  I/DfuImpl: Version number read: 0.8 -> Bootloader from SDK 9.0 or newer. Signature supported
10:42:48.906  W/DfuImpl: Legacy DFU bootloader found
10:42:49.908  I/DfuImpl: Enabling notifications...
10:42:49.911  D/BluetoothGatt: setCharacteristicNotification() - uuid: 00001531-1212-efde-1523-785feabcd123 enable: true
10:42:50.973  I/DfuImpl: Sending Start DFU command (Op Code = 1, Upload Mode = 4)
10:42:51.046  I/DfuImpl: Sending image size array to DFU Packet (0b, 0b, 78820b)
10:42:51.340  I/DfuImpl: Sending the Initialize DFU Parameters START (Op Code = 2, Value = 0)
10:42:51.435  I/DfuImpl: Sending 14 bytes of init packet
10:42:51.436  I/DfuImpl: Sending init packet (Value = 52-00-FF-FF-FF-FF-FF-FF-01-00-B7-00-7E-E8)
10:42:51.444  I/DfuImpl: Sending the Initialize DFU Parameters COMPLETE (Op Code = 2, Value = 1)
10:42:51.532  I/DfuImpl: Sending Receive Firmware Image request (Op Code = 3)
10:42:51.651  I/DfuImpl: Uploading firmware...
10:43:41.698  I/DfuImpl: Response received (Op Code = 16, Req Op Code = 3, Status = 1)
10:43:41.699  I/DfuImpl: Transfer of 78820 bytes has taken 49644 ms
10:43:41.700  I/DfuImpl: Sending Validate request (Op Code = 4)
10:43:41.795  I/DfuImpl: Response received (Op Code = 16, Req Op Code = 4, Status = 1)
10:43:41.805  I/DfuImpl: Sending Activate and Reset request (Op Code = 5)
10:43:42.186  D/BluetoothGatt: onClientConnectionState() - status=19 clientIf=6 device=E0:22:0C:0E:98:BD
10:43:42.187  W/DfuBaseService: Target device disconnected with status: 19
10:43:42.188  D/BluetoothGatt: refresh() - device: E0:22:0C:0E:98:BD
10:43:42.191  I/DfuBaseService: Refreshing result: true
10:43:42.191  I/DfuBaseService: Cleaning up...
10:43:42.192  D/BluetoothGatt: close()
10:43:42.192  D/BluetoothGatt: unregisterApp() - mClientIf=6
10:43:42.194  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_DISCONNECTED
10:43:43.642  I/DfuBaseService: DFU service destroyed
10:43:43.887  D/mali_winsys: new_window_surface returns 0x3000,  [679x88]-format:1
```

Catlog from Android 8.1 (OTA update failed)
``` LOG
10:30:35.150  I/DfuBaseService: DFU service created. Version: 1.9.0
10:30:35.285  I/DfuBaseService: Connecting to the device...
10:30:35.288  D/BluetoothGatt: connect() - device: E0:22:0C:0E:98:BD, auto: false
10:30:35.288  D/BluetoothGatt: registerApp()
10:30:35.288  D/BluetoothGatt: registerApp() - UUID=0e2b0b79-4c6b-43ed-9a43-3cefd31a84d8
10:30:35.291  D/BluetoothGatt: onClientRegistered() - status=0 clientIf=7
10:30:35.345  D/BluetoothGatt: onClientConnectionState() - status=0 clientIf=7 device=E0:22:0C:0E:98:BD
10:30:35.345  I/DfuBaseService: Connected to GATT server
10:30:35.347  I/DfuBaseService: Waiting 1600 ms for a possible Service Changed indication...
10:30:35.347  D/BluetoothGatt: discoverServices() - device: E0:22:0C:0E:98:BD
10:30:35.349  I/DfuBaseService: Attempting to start service discovery... succeed
10:30:35.353  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_CONNECTED
10:30:35.360  D/BluetoothGatt: onSearchComplete() = Device=E0:22:0C:0E:98:BD Status=0
10:30:35.361  I/DfuBaseService: Services discovered
10:30:35.792  D/BluetoothGatt: onConnectionUpdated() - Device=E0:22:0C:0E:98:BD interval=9 latency=0 timeout=200 status=0
10:30:36.386  I/DfuImpl: Reading DFU version number...
10:30:36.413  I/DfuImpl: Version number read: 0.1 -> Application with Legacy buttonless update from SDK 7.0 or newer
10:30:36.417  I/DfuImpl: Reading Service Changed CCCD value...
10:30:36.447  W/DfuImpl: Application with legacy buttonless update found
10:30:36.449  I/DfuImpl: Enabling notifications...
10:30:36.452  D/BluetoothGatt: setCharacteristicNotification() - uuid: 00001531-1212-efde-1523-785feabcd123 enable: true
10:30:37.504  I/DfuImpl: Sending Start DFU command (Op Code = 1, Upload Mode = 4)
10:30:39.562  D/BluetoothGatt: onClientConnectionState() - status=8 clientIf=7 device=E0:22:0C:0E:98:BD
10:30:39.563  W/DfuBaseService: Target device disconnected with status: 8
10:30:39.566  I/DfuBaseService: Cleaning up...
10:30:39.567  D/BluetoothGatt: close()
10:30:39.567  D/BluetoothGatt: unregisterApp() - mClientIf=7
10:30:39.571  I/DfuImpl: Starting service that will connect to the DFU bootloader
10:30:39.574  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_DISCONNECTED
10:30:41.631  I/DfuBaseService: Connecting to the device...
10:30:41.633  D/BluetoothGatt: connect() - device: E0:22:0C:0E:98:BD, auto: false
10:30:41.633  D/BluetoothGatt: registerApp()
10:30:41.634  D/BluetoothGatt: registerApp() - UUID=657996be-c909-481a-9321-a7fe8c78efb9
10:30:41.637  D/BluetoothGatt: onClientRegistered() - status=0 clientIf=7
10:30:41.871  D/BluetoothGatt: onClientConnectionState() - status=0 clientIf=7 device=E0:22:0C:0E:98:BD
10:30:41.872  I/DfuBaseService: Connected to GATT server
10:30:41.875  I/DfuBaseService: Waiting 1600 ms for a possible Service Changed indication...
10:30:41.876  D/BluetoothGatt: discoverServices() - device: E0:22:0C:0E:98:BD
10:30:41.878  I/DfuBaseService: Attempting to start service discovery... succeed
10:30:41.879  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_CONNECTED
10:30:41.898  D/BluetoothGatt: onSearchComplete() = Device=E0:22:0C:0E:98:BD Status=0
10:30:41.899  I/DfuBaseService: Services discovered
10:30:42.917  I/DfuImpl: Reading DFU version number...
10:30:47.139  E/DfuImpl: Characteristic read error: 133
10:30:47.140  D/BluetoothGatt: onClientConnectionState() - status=8 clientIf=7 device=E0:22:0C:0E:98:BD
10:30:47.140  W/DfuBaseService: Target device disconnected with status: 8
10:30:47.141  E/DfuBaseService: Unable to read version number: device disconnected
10:30:47.141  I/DfuBaseService: Cleaning up...
10:30:47.141  D/BluetoothGatt: close()
10:30:47.142  D/BluetoothGatt: unregisterApp() - mClientIf=7
10:30:47.204  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_DISCONNECTED
10:30:47.224  I/DfuBaseService: DFU service destroyed
```

**Comment by beegee-tokyo** [comment]
Additional info. I tried to put the device into bootloader OTA mode manually and then start the DFU update.
Then the nRF52 is blinking fast (receiving data), but catlog shows OPERATION FAILED and gets stuck there.
Catlog from Android 8.1:
``` LOG
16:39:18.852  I/DfuBaseService: DFU service created. Version: 1.9.0
16:39:18.984  I/DfuBaseService: Connecting to the device...
16:39:18.985  D/BluetoothGatt: connect() - device: E0:22:0C:0E:98:BE, auto: false
16:39:18.985  D/BluetoothGatt: registerApp()
16:39:18.986  D/BluetoothGatt: registerApp() - UUID=2af9102f-126f-429f-bba1-e84c08b8240e
16:39:18.988  D/BluetoothGatt: onClientRegistered() - status=0 clientIf=7
16:39:19.213  D/BluetoothGatt: onClientConnectionState() - status=0 clientIf=7 device=E0:22:0C:0E:98:BE
16:39:19.214  I/DfuBaseService: Connected to GATT server
16:39:19.219  D/BluetoothGatt: discoverServices() - device: E0:22:0C:0E:98:BE
16:39:19.221  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_CONNECTED
16:39:19.225  I/DfuBaseService: Attempting to start service discovery... succeed
16:39:19.522  D/BluetoothGatt: onConnectionUpdated() - Device=E0:22:0C:0E:98:BE interval=6 latency=0 timeout=500 status=0
16:39:19.833  D/BluetoothGatt: onSearchComplete() = Device=E0:22:0C:0E:98:BE Status=0
16:39:19.833  I/DfuBaseService: Services discovered
16:39:19.881  D/BluetoothGatt: onConnectionUpdated() - Device=E0:22:0C:0E:98:BE interval=36 latency=0 timeout=500 status=0
16:39:20.865  I/DfuImpl: Reading DFU version number...
16:39:21.008  I/DfuImpl: Version number read: 0.8 -> Bootloader from SDK 9.0 or newer. Signature supported
16:39:21.013  W/DfuImpl: Legacy DFU bootloader found
16:39:21.862  I/zygote: Do partial code cache collection, code=504KB, data=355KB
16:39:21.864  I/zygote: After code cache collection, code=504KB, data=355KB
16:39:21.864  I/zygote: Increasing code cache capacity to 2MB
16:39:22.015  I/DfuImpl: Enabling notifications...
16:39:22.016  D/BluetoothGatt: setCharacteristicNotification() - uuid: 00001531-1212-efde-1523-785feabcd123 enable: true
16:39:23.134  I/DfuImpl: Sending Start DFU command (Op Code = 1, Upload Mode = 4)
16:39:23.258  I/DfuImpl: Sending image size array to DFU Packet (0b, 0b, 78820b)
16:39:23.709  I/DfuImpl: Sending the Initialize DFU Parameters START (Op Code = 2, Value = 0)
16:39:23.797  I/DfuImpl: Sending 14 bytes of init packet
16:39:23.798  I/DfuImpl: Sending init packet (Value = 52-00-FF-FF-FF-FF-FF-FF-01-00-B7-00-7E-E8)
16:39:23.804  I/DfuImpl: Sending the Initialize DFU Parameters COMPLETE (Op Code = 2, Value = 1)
16:39:23.889  I/DfuImpl: Sending the number of packets before notifications (Op Code = 8, Value = 12)
16:39:24.024  I/DfuImpl: Sending Receive Firmware Image request (Op Code = 3)
16:39:24.130  I/DfuImpl: Uploading firmware...
16:39:24.162  I/DfuImpl: Response received (Op Code = 16, Req Op Code = 3, Status = 6)
16:39:24.163  E/DfuImpl: Device returned error after sending file (error 6): OPERATION FAILED
16:39:24.164 _android I/DfuImpl: Sending Reset command (Op Code = 6)
```
Nothing is appearing in the log after the "Sending Reset command".

On Android 7.1.1 this procedure works perfect:
``` LOG
16:51:28.095  I/DfuBaseService: DFU service created. Version: 1.9.0
16:51:28.180  I/DfuBaseService: Connecting to the device...
16:51:28.181  D/BluetoothGatt: connect() - device: E0:22:0C:0E:98:BE, auto: false
16:51:28.181  D/BluetoothGatt: registerApp()
16:51:28.182  D/BluetoothGatt: registerApp() - UUID=49600036-cc7a-40b9-a4f3-38ef09858c6b
16:51:28.183  D/BluetoothGatt: onClientRegistered() - status=0 clientIf=6
16:51:32.893  D/BluetoothGatt: onClientConnectionState() - status=0 clientIf=6 device=E0:22:0C:0E:98:BE
16:51:32.894  I/DfuBaseService: Connected to GATT server
16:51:32.896  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_CONNECTED
16:51:32.897  D/BluetoothGatt: discoverServices() - device: E0:22:0C:0E:98:BE
16:51:32.898  I/DfuBaseService: Attempting to start service discovery... succeed
16:51:33.549  D/BluetoothGatt: onSearchComplete() = Device=E0:22:0C:0E:98:BE Status=0
16:51:33.549  I/DfuBaseService: Services discovered
16:51:34.565  I/DfuImpl: Reading DFU version number...
16:51:34.750  W/BluetoothGatt: onCharacteristicRead() - Device=E0:22:0C:0E:98:BE handle=21 Status=0
16:51:34.751  I/DfuImpl: Version number read: 0.8 -> Bootloader from SDK 9.0 or newer. Signature supported
16:51:34.753  W/DfuImpl: Legacy DFU bootloader found
16:51:35.754  I/DfuImpl: Enabling notifications...
16:51:35.754  D/BluetoothGatt: setCharacteristicNotification() - uuid: 00001531-1212-efde-1523-785feabcd123 enable: true
16:51:36.825  I/DfuImpl: Sending Start DFU command (Op Code = 1, Upload Mode = 4)
16:51:36.895  I/DfuImpl: Sending image size array to DFU Packet (0b, 0b, 64628b)
16:51:37.188  I/DfuImpl: Sending the Initialize DFU Parameters START (Op Code = 2, Value = 0)
16:51:37.287  I/DfuImpl: Sending 14 bytes of init packet
16:51:37.287  I/DfuImpl: Sending init packet (Value = 52-00-FF-FF-FF-FF-FF-FF-01-00-B7-00-CD-2E)
16:51:37.292  I/DfuImpl: Sending the Initialize DFU Parameters COMPLETE (Op Code = 2, Value = 1)
16:51:37.385  I/DfuImpl: Sending Receive Firmware Image request (Op Code = 3)
16:51:37.490  I/DfuImpl: Uploading firmware...
16:52:18.482  I/DfuImpl: Response received (Op Code = 16, Req Op Code = 3, Status = 1)
16:52:18.483  I/DfuImpl: Transfer of 64628 bytes has taken 40484 ms
16:52:18.483  I/DfuImpl: Sending Validate request (Op Code = 4)
16:52:18.580  I/DfuImpl: Response received (Op Code = 16, Req Op Code = 4, Status = 1)
16:52:18.597  I/DfuImpl: Sending Activate and Reset request (Op Code = 5)
16:52:18.971  D/BluetoothGatt: onClientConnectionState() - status=19 clientIf=6 device=E0:22:0C:0E:98:BE
16:52:18.971  W/DfuBaseService: Target device disconnected with status: 19
16:52:18.973  D/BluetoothGatt: refresh() - device: E0:22:0C:0E:98:BE
16:52:18.975  I/DfuBaseService: Refreshing result: true
16:52:18.975  I/DfuBaseService: Cleaning up...
16:52:18.976  D/BluetoothGatt: close()
16:52:18.976  D/BluetoothGatt: unregisterApp() - mClientIf=6
16:52:18.981  I/DfuBaseService: Action received: android.bluetooth.device.action.ACL_DISCONNECTED
16:52:20.413  I/DfuBaseService: DFU service destroyed
16:52:20.653  D/mali_winsys: new_window_surface returns 0x3000,  [679x88]-format:1
```

**Comment by beegee-tokyo** [comment]
Tried with nRFConnect application and same as nRFToolbox. DFU update fails on Android 8 and 8.1
Log from nRFConnect:
```
nRF Connect, 2019-03-24
PPG-04E1E1CD2CACC007 (E0:22:0C:0E:98:BD)
V	10:00:14.734	[DFU] DFU service started
V	10:00:14.735	[DFU] Opening file...
I	10:00:14.788	[DFU] Firmware file opened successfully
V	10:00:14.788	[DFU] Connecting to DFU target...
D	10:00:14.792	[DFU] gatt = device.connectGatt(autoConnect = false)
I	10:00:14.800	[DFU] Connected to E0:22:0C:0E:98:BD
D	10:00:14.801	[DFU] wait(1600)
V	10:00:16.402	[DFU] Discovering services...
D	10:00:16.403	[DFU] gatt.discoverServices()
I	10:00:16.412	[DFU] Services discovered
D	10:00:16.458	[DFU] wait(1000)
V	10:00:17.456	[DFU] Reading DFU version number...
D	10:00:17.456	[DFU] gatt.readCharacteristic(00001534-1212-efde-1523-785feabcd123)
I	10:00:17.456	[DFU] Read Response received from 00001534-1212-efde-1523-785feabcd123, value (0x): 01-00
A	10:00:17.456	[DFU] Version number read: 0.1
V	10:00:17.456	[DFU] Reading Service Changed CCCD value...
D	10:00:17.456	[DFU] gatt.readDescriptor(00002902-0000-1000-8000-00805f9b34fb)
I	10:00:17.475	[DFU] Read Response received from descr.00002a05-0000-1000-8000-00805f9b34fb, value (0x): 02-00
A	10:00:17.476	[DFU] Service Changed indications enabled
W	10:00:17.476	[DFU] Application with buttonless update found
V	10:00:17.476	[DFU] Jumping to the DFU Bootloader...
V	10:00:17.477	[DFU] Enabling notifications for 00001531-1212-efde-1523-785feabcd123
D	10:00:17.478	[DFU] gatt.setCharacteristicNotification(00001531-1212-efde-1523-785feabcd123, true)
D	10:00:17.481	[DFU] gatt.writeDescriptor(00002902-0000-1000-8000-00805f9b34fb, value=0x01-00)
I	10:00:17.509	[DFU] Data written to descr.00001531-1212-efde-1523-785feabcd123, value (0x): 01-00
V	10:00:17.509	[DFU] Notifications enabled for 00001531-1212-efde-1523-785feabcd123
A	10:00:17.510	[DFU] Notifications enabled
D	10:00:17.510	[DFU] wait(1000)
V	10:00:18.532	[DFU] Writing to characteristic 00001531-1212-efde-1523-785feabcd123
D	10:00:18.533	[DFU] gatt.writeCharacteristic(00001531-1212-efde-1523-785feabcd123)
D	10:00:20.534	[Callback] Connection state changed with status: 8 and new state: DISCONNECTED (0)
E	10:00:20.535	Error 8 (0x8): GATT CONN TIMEOUT
I	10:00:20.536	Disconnected
A	10:00:20.537	[DFU] Jump to bootloader sent (Op Code = 1, Upload Mode = 4)
I	10:00:20.611	[DFU] Disconnected by the remote device
D	10:00:20.611	[DFU] gatt.close()
D	10:00:20.612	[Broadcast] Action received: android.bluetooth.device.action.ACL_DISCONNECTED
D	10:00:20.628	[DFU] [Broadcast] Action received: android.bluetooth.device.action.ACL_DISCONNECTED
V	10:00:20.629	[DFU] DFU service started
I	10:00:20.629	[DFU] Firmware file opened successfully
D	10:00:20.629	[DFU] wait(1000)
D	10:00:21.618	[DFU] wait(1000)
V	10:00:22.619	[DFU] Connecting to DFU target...
D	10:00:22.683	[DFU] gatt = device.connectGatt(autoConnect = false)
I	10:00:22.856	[DFU] Connected to E0:22:0C:0E:98:BD
D	10:00:22.864	[Broadcast] Action received: android.bluetooth.device.action.ACL_CONNECTED
D	10:00:22.865	[DFU] wait(1600)
D	10:00:22.865	[DFU] [Broadcast] Action received: android.bluetooth.device.action.ACL_CONNECTED
V	10:00:24.462	[DFU] Discovering services...
D	10:00:24.462	[DFU] gatt.discoverServices()
I	10:00:24.472	[DFU] Services discovered
D	10:00:24.503	[DFU] wait(1000)
V	10:00:25.512	[DFU] Reading DFU version number...
D	10:00:25.512	[DFU] gatt.readCharacteristic(00001534-1212-efde-1523-785feabcd123)
E	10:00:28.173	[DFU] Error (0x85): GATT ERROR
D	10:00:28.173	[DFU] gatt.close()
D	10:00:28.173	[DFU] wait(600)
D	10:00:28.188	[Broadcast] Action received: android.bluetooth.device.action.ACL_DISCONNECTED
D	10:00:28.204	[DFU] [Broadcast] Action received: android.bluetooth.device.action.ACL_DISCONNECTED
D	10:00:28.778	gatt.close()
D	10:00:28.780	wait(200)
V	10:00:28.981	Connecting to E0:22:0C:0E:98:BD...
D	10:00:28.982	gatt = device.connectGatt(autoConnect = false, TRANSPORT_LE, preferred PHY = LE 1M)
D	10:00:29.066	[Callback] Connection state changed with status: 0 and new state: CONNECTED (2)
I	10:00:29.066	Connected to E0:22:0C:0E:98:BD
D	10:00:29.067	wait(1600ms)
D	10:00:29.117	[Broadcast] Action received: android.bluetooth.device.action.ACL_CONNECTED
I	10:00:29.469	Connection parameters updated (interval: 11.25ms, latency: 0, timeout: 2000ms)
V	10:00:30.670	Discovering services...
D	10:00:30.670	gatt.discoverServices()
D	10:00:30.695	[Callback] Services discovered with status: 0
I	10:00:30.695	Services discovered
V	10:00:30.741	Generic Access (0x1800)
- Device Name [R W] (0x2A00)
- Appearance [R] (0x2A01)
- Peripheral Preferred Connection Parameters [R] (0x2A04)
- Central Address Resolution [R] (0x2AA6)
Generic Attribute (0x1801)
- Service Changed [I] (0x2A05)
Client Characteristic Configuration (0x2902)
Device Firmware Update Service (00001530-1212-efde-1523-785feabcd123)
- DFU Packet [WNR] (00001532-1212-efde-1523-785feabcd123)
- DFU Control Point [N W] (00001531-1212-efde-1523-785feabcd123)
Client Characteristic Configuration (0x2902)
- DFU Version [R] (00001534-1212-efde-1523-785feabcd123)
Device Information (0x180A)
- Model Number String [R] (0x2A24)
- Serial Number String [R] (0x2A25)
- Firmware Revision String [R] (0x2A26)
- Hardware Revision String [R] (0x2A27)
- Software Revision String [R] (0x2A28)
- Manufacturer Name String [R] (0x2A29)
Unknown Service (0000f0f0-0000-1000-8000-00805f9b34fb)
- Unknown Characteristic [N W] (0000f0f1-0000-1000-8000-00805f9b34fb)
Client Characteristic Configuration (0x2902)
Unknown Service (0000f0e0-0000-1000-8000-00805f9b34fb)
- Unknown Characteristic [R W] (0000f0e1-0000-1000-8000-00805f9b34fb)
D	10:00:30.743	gatt.setCharacteristicNotification(00002a05-0000-1000-8000-00805f9b34fb, true)
D	10:00:30.746	gatt.setCharacteristicNotification(0000f0f1-0000-1000-8000-00805f9b34fb, true)
```

**Comment by beegee-tokyo** [comment]
Update:
DFU on Android 8.0 and 8.1 only fails if the device is paired/bonded. 
If I unpair / remove bond of the device, the DFU update works on both Android 8.0 and 8.1.

I think that it might be necessary to add a third step to the DFU sequence on Android >= 8 (SDK >= 26)
1) after device is selected, check if it is bonded. If bonded, remove bond
2) reconnect to unpaired device and send Start DFU command
3) reconnect to device and start DFU

For now my workaround is to advice the user to unpair the device if it is paired/bonded. My application doesn't need pairing to work so that is a temporary solution for me.

Additional, I had to set PRN to 4 to make it work.
The default PRN is set to 12, but in LegacyDfuImpl.java line 455 it says
```//       It has been tested that PRN = 10 may be the highest supported value.```
which is inconsistent with the default value.
Maybe on Android 8.0 and newer the PRN default should be set to 4 and on older SDK the default should be set to 10 instead of 12.

**Comment by beegee-tokyo** [comment]
Last update and closing this issue. Hope you can confirm the solution and implement these changes into the DFU library.

Solved the problem with following change in DfuBaseService:

In 
**`DfuBaseService.java >> private final BluetoothGattCallback mGattCallback = new BluetoothGattCallback() >> public void onConnectionStateChange()`**
I changed
```JAVA
if (gatt.getDevice().getBondState() == BluetoothDevice.BOND_BONDED) {
	logi("Waiting 1600 ms for a possible Service Changed indication...");
	waitFor(1600);
	// After 1.6s the services are already discovered so the following gatt.discoverServices() finishes almost immediately.

	// NOTE: This also works with shorted waiting time. The gatt.discoverServices() must be called after the indication is received which is
	// about 600ms after establishing connection. Values 600 - 1600ms should be OK.
}
```
to
```JAVA
if (gatt.getDevice().getBondState() == BluetoothDevice.BOND_BONDED) {
	logi("Remove bond before starting DFU");
	try {
		Method m = gatt.getDevice().getClass()
			.getMethod("removeBond", (Class[]) null);
		m.invoke(gatt.getDevice(), (Object[]) null);
	} catch (Exception e) {
		Log.e(TAG, e.getMessage());
	}
	return;
}
```

That solved the DFU problem on Android 8.0 and 8.1 and works as well on Android 7.1.1 and older (tested on Samsung Galaxy Tab 10 Android 7.1.1 and Samsung Galaxy Tab A Android 5.0.2).

In addition I limit PRN with
```JAVA
if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.O) {
//	starter.setPacketsReceiptNotificationsValue(DfuServiceInitiator.DEFAULT_PRN_VALUE);
	starter.setPacketsReceiptNotificationsValue(10);
	Log.e(TAG, "Android SDK < 26 (8.0) set PRN to 10");
} else {
	starter.setPacketsReceiptNotificationsValue(4);
	Log.e(TAG, "Android SDK >= 26 (8.0) set PRN to 4");
}
```
when calling the DfuServiceInitiator.

**Comment by beegee-tokyo** [comment]
@philips77 
Just to push a notification to you about above solution.
